src/x/server/server.go

```diff
@@ -140,11 +156,78 @@ func (s *server) Serve(l net.Listener) error {
 	return nil
 }
 
+func (s *server) isTLSConfigCacheValid() bool {
+	if s.tlsConfigCache == nil || s.certsTTLTicker == nil {
+		return false
+	}
+	select {
+	case <-s.certsTTLTicker.C:
+		return false
+	default:
+		return true
+	}
+}
+
+func (s *server) getTLSConfig() (*tls.Config, error) {
+	s.Lock()
```

Comment:
Is this exclusive lock called on every request? That seems like it could be a lot more lock contention than needed (albeit hopefully for a short critical path). 

Can we use an `RWLock` here instead?